music/src/piano_genie/model.ts

```diff
@@ -0,0 +1,295 @@
+      throw new Error('Invalid previous output specified.');
+    }
+    if (deltaTime < 0) {
+      throw new Error('Negative delta time specified.');
+    }
+
+    // Ensure that the model is initialized.
+    if (!this.initialized) {
+      throw new Error('Model is not initialized.');
+    }
+
+    // Compute logits.
+    const [finalState, decLogits]: [LSTMState, tf.Tensor1D] = tf.tidy(() => {
+      // Initialize decoder feats array.
+      const decFeatsArr: tf.Tensor2D[] = [];
+
+      // Add button input to decoder feats and translate to [-1, 1].
+      const buttonTensor = tf.tensor2d([button], [1, 1], 'float32');
+      const buttonScaled = 
+        tf.sub(tf.mul(2., tf.div(buttonTensor, NBUTTONS - 1)), 1);
+      decFeatsArr.push(buttonScaled as tf.Tensor2D);
+
+      // Add autoregression (history) to decoder feats.
+      const lastOutputTensor = tf.tensor1d([lastOutput], 'int32');
+      const lastOutputInc = 
+        tf.add(lastOutputTensor, tf.scalar(1, 'int32')) as tf.Tensor1D;
+      const lastOutputOh = 
+        tf.cast(tf.oneHot(lastOutputInc, NPIANOKEYS + 1), 'float32');
+      decFeatsArr.push(lastOutputOh);
+
+      // Add delta times to decoder feats.
+      const deltaTimeTensor = tf.tensor1d([deltaTime], 'float32');
+      const deltaTimeBin = 
+        tf.round(tf.mul(deltaTimeTensor, DATA_TIME_QUANTIZE_RATE));
+      const deltaTimeTrunc = tf.minimum(deltaTimeBin, DATA_MAX_DISCRETE_TIMES);
+      const deltaTimeInt = 
+        tf.cast(tf.add(deltaTimeTrunc, 1e-4), 'int32') as tf.Tensor1D;
+      const deltaTimeOh = tf.oneHot(deltaTimeInt, DATA_MAX_DISCRETE_TIMES + 1);
+      const deltaTimeOhFloat = tf.cast(deltaTimeOh, 'float32');
+      decFeatsArr.push(deltaTimeOhFloat);
+
+      // Project feats array through RNN input matrix.
+      let rnnInput: tf.Tensor2D = tf.concat(decFeatsArr, 1);
+      rnnInput = tf.matMul(
+        rnnInput,
+        this.modelVars[
+          'phero_model/decoder/rnn_input/dense/kernel'] as tf.Tensor2D);
+      rnnInput = tf.add(
+        rnnInput,
+        this.modelVars[
+          'phero_model/decoder/rnn_input/dense/bias'] as tf.Tensor1D);
+
+      // Evaluate RNN.
+      const [c, h] = tf.multiRNNCell(
+        this.decLSTMCells, rnnInput, this.state.c, this.state.h);
+      const finalState: LSTMState = { c, h };
+
+      // Project to logits.
+      let decLogits: tf.Tensor2D = tf.matMul(
+        h[RNN_NLAYERS - 1],
+        this.modelVars[
+          'phero_model/decoder/pitches/dense/kernel'] as tf.Tensor2D);
+      decLogits = tf.add(
+        decLogits,
+        this.modelVars[
+          'phero_model/decoder/pitches/dense/bias'] as tf.Tensor1D);
+
+      // Remove batch axis to produce piano key (n=88) logits.
+      const decLogits1D = tf.reshape(decLogits, [NPIANOKEYS]) as tf.Tensor1D;
+
+      return [finalState, decLogits1D] as [LSTMState, tf.Tensor1D];
+    });
+
+    disposeState(this.state);
+    this.state = finalState;
+
+    return decLogits;
```

Comment:
same comment as before, I do not think we should be returning tensors, and definitely not logits, to the user